File: sdk/java/src/main/java/io/ancora/RunHandle.java
package io.ancora;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.NoSuchElementException;

public final class RunHandle {

    private final Runtime runtime;
    private final String runId;

    RunHandle(Runtime runtime, String runId) {
        this.runtime = runtime;
        this.runId = runId;
    }

    public String runId() {
        return runId;
    }

    public Iterable<RunEvent> events() {
        return () -> new Iterator<>() {
            private RunEvent pending = null;
            private boolean done = false;

            @Override
            public boolean hasNext() {
                if (done) return false;
                if (pending != null) return true;
                try {
                    String json = runtime.pollEvent(runId);
                    if (json == null) {
                        done = true;
                        return false;
                    }
                    pending = Wire.parseEvent(json);
                    return true;
                } catch (Throwable t) {
                    throw new RuntimeException("Error polling next event", t);
                }
            }

            @Override
            public RunEvent next() {
                if (!hasNext()) throw new NoSuchElementException();
                RunEvent e = pending;
                pending = null;
                return e;
            }
        };
    }

    public List<RunEvent> collectAll() {
        List<RunEvent> list = new ArrayList<>();
        for (RunEvent e : events()) list.add(e);
        return list;
    }

    public RunHandle resume(byte[] decisionBytes) throws Throwable {
        runtime.resumeRun(runId, decisionBytes);
        return this;
    }

    public RunHandle resume(String decisionJson) throws Throwable {
        return resume(decisionJson.getBytes(StandardCharsets.UTF_8));
    }

    public List<RunEvent> resumeAndCollectAll(byte[] decisionBytes) throws Throwable {
        resume(decisionBytes);
        return collectAll();
    }

    public String getCost() throws Throwable {
        return runtime.getCost(runId);
    }
}

File: sdk/dotnet/Ancora/RunHandle.cs
using System;
using System.Collections.Generic;
using System.Runtime.CompilerServices;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace Ancora;

/// <summary>
/// A handle to a live or completed run.
/// Provides an IAsyncEnumerable event stream and resume support.
/// </summary>
public sealed class RunHandle
{
    /// <summary>
    /// The stable run ID assigned when the run was created.
    /// </summary>
    public string RunId { get; }

    private readonly Runtime _runtime;

    internal RunHandle(string runId, Runtime runtime)
    {
        RunId = runId ?? throw new ArgumentNullException(nameof(runId));
        _runtime = runtime ?? throw new ArgumentNullException(nameof(runtime));
    }

    /// <summary>
    /// Stream events from this run as an async sequence.
    /// The stream ends when the run produces no further events.
    /// </summary>
    public async IAsyncEnumerable<RunEvent> EventsAsync(
        [EnumeratorCancellation] CancellationToken cancellationToken = default)
    {
        while (!cancellationToken.IsCancellationRequested)
        {
            var raw = _runtime.PollEvent(RunId);
            if (raw is null) yield break;
            yield return Wire.ParseEvent(raw);
            await Task.Yield();
        }
    }

    /// <summary>
    /// Collect all remaining events from this run into a list.
    /// Returns after the run produces no further events.
    /// </summary>
    public async Task<IReadOnlyList<RunEvent>> CollectAsync(
        CancellationToken cancellationToken = default)
    {
        var events = new List<RunEvent>();
        await foreach (var ev in EventsAsync(cancellationToken).WithCancellation(cancellationToken))
            events.Add(ev);
        return events;
    }

    /// <summary>
    /// Resume a suspended run, injecting a decision payload as a UTF-8 string.
    /// </summary>
    public void Resume(string decision)
    {
        ArgumentNullException.ThrowIfNull(decision);
        var bytes = Wire.EncodeDecision(decision);
        _runtime.ResumeRun(RunId, bytes);
    }

    /// <summary>
    /// Resume a suspended run with a raw byte decision payload.
    /// </summary>
    public void Resume(ReadOnlySpan<byte> decisionBytes) =>
        _runtime.ResumeRun(RunId, decisionBytes);

    /// <summary>
    /// Resume a suspended run and then collect all subsequent events.
    /// </summary>
    public async Task<IReadOnlyList<RunEvent>> ResumeAndCollectAsync(
        string decision,
        CancellationToken cancellationToken = default)
    {
        Resume(decision);
        return await CollectAsync(cancellationToken);
    }

    /// <summary>
    /// Return the cost summary JSON for this run.
    /// </summary>
    public string GetCost() => _runtime.GetCost(RunId);
}
